### PR TITLE
feat: pricing module + category improvements

### PR DESCRIPTION
**Pull request by VariableVic**
### Pricing module
- Added pricing module support to the following routes:
  - `/app/api/categories/[...handle]/route.ts`
  - `/app/api/collections/[handle]/route.ts`
  - `/app/api/products/[handle]/route.ts`
  - `/app/api/collections/route.ts`

### Categories
- Refactored the category routes with [catch-all segments](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes#catch-all-segments).

### CODE REVIEW

File: src/app/api/categories/[...handle]/route.ts [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        id
+        title
+        handle
+        tags
+        status
+        collection
+        collection_id
+        thumbnail
+        images {
+            url
```
Comment (riqwan) [inline src/app/api/categories/[...handle]/route.ts:115]: is the indentation 4 spaces on purpose?

File: src/app/api/categories/[...handle]/route.ts [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        (price) => price.currency_code === currency_code
+      )
+
+      if (!price) continue
+
+      delete variant.price
+      variant.price = price
+      variant.calculated_price = price.amount
+    }
+  }
```
Comment (riqwan) [inline src/app/api/categories/[...handle]/route.ts:173]:
```suggestion
      const [price] = await modules.pricingService.calculatePrices(
        { id: priceSetIds },
        {
          context: { currency_code },
        }
      )

      delete variant.price
      
      if (!price) continue

      variant.price = price
      variant.calculated_price = price.amount
    }
  }
```

File: src/app/api/collections/[handle]/route.ts [hunk 3]
```diff
@@ -34,40 +35,135 @@ export async function GET(
...
+
+      const price = prices.find(
+        (price) => price.currency_code === currency_code
+      )
+
+      if (!price) continue
+
+      delete variant.price
+      variant.price = price
+      variant.calculated_price = price.amount
```
Comment (riqwan) [inline src/app/api/collections/[handle]/route.ts:161]:
```suggestion
      const [price] = await modules.pricingService.calculatePrices(
        { id: priceSetIds },
        {
          context: { currency_code },
        }
      )

      delete variant.price
      
      if (!price) continue

      variant.price = price
      variant.calculated_price = price.amount
```

File: src/app/api/products/[handle]/route.ts [hunk 4]
```diff
@@ -36,7 +35,6 @@ export async function GET(
     }
   )
 
-  const productsWithPrices = await getPrices(data, cart_id, region_id)
-
-  return NextResponse.json({ products: productsWithPrices })
+  // Return the response
+  return NextResponse.json({ products })
```
Comment (riqwan) [inline src/app/api/products/[handle]/route.ts:39]: prices not needed here?

File: src/app/api/products/route.ts [hunk 5]
```diff
@@ -38,75 +20,128 @@ export async function GET(request: NextRequest) {
...
+        (price) => price.currency_code === currency_code
+      )
+
+      if (!price) continue
+
+      delete variant.price
+      variant.price = price
+      variant.calculated_price = price.amount
+    }
+  }
```
Comment (riqwan) [inline src/app/api/products/route.ts:136]: suggestion: you can extract this out into a util

File: src/lib/data/index.ts [hunk 6]
```diff
@@ -238,17 +267,84 @@ export async function getProductsByCollectionHandle({
...
+ */
+export async function getCategoriesList(
+  offset: number = 0,
+  limit?: number
+): Promise<{
+  product_categories: ProductCategoryWithChildren[]
+  count: number
+}> {
+  if (PRODUCT_MODULE_ENABLED) {
+    DEBUG && console.log("PRODUCT_MODULE_ENABLED")
```
Comment (riqwan) [inline src/lib/data/index.ts:286]: missed console?

**Review by riqwan — APPROVED** [review]
added some todos and suggestions, but overall lgtm.

File: src/app/api/categories/[...handle]/route.ts [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+  const { page, limit } = searchParams
+
+  let { handle: categoryHandle } = params
+
+  const handle = categoryHandle.map((handle: string, index: number) =>
+    categoryHandle.slice(0, index + 1).join("/")
+  )
+
+  const product_categories = await productService
+    .listCategories(
```
[the remark this answers is not in the document]
Comment (VariableVic) [reply src/app/api/categories/[...handle]/route.ts:28]: Does `retrieveCategory` work with a category handle?

[the remark this answers is not in the document]
Comment (riqwan) [reply src/app/api/categories/[...handle]/route.ts:28]: ah, fair point. No it doesn't.

File: src/app/api/categories/[...handle]/route.ts (hunk 2 above)
Comment (VariableVic) [reply src/app/api/categories/[...handle]/route.ts:173]: This doesn't work as expected if the product has multiple priceSetIds. [price] on line 156 will return the first array item, but that's not necessarily the one that matches the currency code. That's why I added the .find, to grab the price that matches the requested currency code. Or am I missing something here?

Comment (VariableVic) [reply src/app/api/categories/[...handle]/route.ts:173]: pricingService.calculatePrices is throwing a TS error, might be related:

```
Property 'calculatePrices' does not exist on type '{ __joinerConfig: ModuleJoinerConfig; __definition: ModuleDefinition; } | { __joinerConfig: ModuleJoinerConfig; __definition: ModuleDefinition; }[]'.
  Property 'calculatePrices' does not exist on type '{ __joinerConfig: ModuleJoinerConfig; __definition: ModuleDefinition; }'.
```

File: next.config.js [hunk 8]
```diff
@@ -4,7 +4,10 @@ const store = require("./store.config.json")
 module.exports = withStoreConfig({
   experimental: {
     serverActions: true,
-    serverComponentsExternalPackages: ["@medusajs/product"],
+    serverComponentsExternalPackages: [
+      "@medusajs/product",
```
[the remark this answers is not in the document]
Comment (VariableVic) [reply next.config.js:8]: Looks like it's still needed, build breaks without it

File: src/app/api/categories/[...handle]/route.ts (hunk 1 above)
Comment (VariableVic) [reply src/app/api/categories/[...handle]/route.ts:127]: nope, fixed

**Pull request merged**